pytests/test_types.py

```diff
@@ -0,0 +1,541 @@
+    )
+
+
+def test_energy_to_pb() -> None:
+    """Test the client energy type conversions to protobuf."""
+    assert_conversion_to_pb(
+        original=Energy(mwh=Decimal("5")),
+        expected_pb=energy_pb2.Energy(mwh=decimal_pb2.Decimal(value="5")),
+        assert_func=assert_equal,
+    )
+
+
+def test_energy_from_pb() -> None:
+    """Test the client energy type conversions from protobuf."""
+    assert_conversion_from_pb(
+        original_pb=energy_pb2.Energy(mwh=decimal_pb2.Decimal(value="5")),
+        expected=Energy(mwh=Decimal("5")),
+        assert_func=assert_equal,
+    )
+
+
+def test_delivery_duration_to_pb() -> None:
+    """Test the client delivery duration type conversions to protobuf."""
+    assert_conversion_to_pb(
+        original=DeliveryDuration.MINUTES_15,
+        expected_pb=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+        assert_func=assert_equal,
+    )
+
+
+def test_delivery_duration_from_pb() -> None:
+    """Test the client delivery duration type conversions from protobuf."""
+    assert_conversion_from_pb(
+        original_pb=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+        expected=DeliveryDuration.MINUTES_15,
+        assert_func=assert_equal,
+    )
+
+
+def test_delivery_period_to_pb() -> None:
+    """Test the client delivery period type conversions to protobuf."""
+    assert_conversion_to_pb(
+        original=DeliveryPeriod(start=START_TIME, duration=timedelta(minutes=15)),
+        expected_pb=delivery_duration_pb2.DeliveryPeriod(
+            start=START_TIME_PB,
+            duration=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+        ),
+        assert_func=assert_equal,
+    )
+
+
+def test_delivery_period_from_pb() -> None:
+    """Test the client delivery period type conversions from protobuf."""
+    assert_conversion_from_pb(
+        original_pb=delivery_duration_pb2.DeliveryPeriod(
+            start=START_TIME_PB,
+            duration=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+        ),
+        expected=DeliveryPeriod(start=START_TIME, duration=timedelta(minutes=15)),
+        assert_func=assert_equal,
+    )
+
+
+def test_invalid_duration_raises_value_error() -> None:
+    """Test that an invalid duration raises a ValueError."""
+    with pytest.raises(ValueError):
+        DeliveryPeriod(start=datetime.now(timezone.utc), duration=timedelta(minutes=10))
+
+
+def test_no_timezone_raises_value_error() -> None:
+    """Test that a datetime without timezone raises a ValueError."""
+    with pytest.raises(ValueError):
+        DeliveryPeriod(start=datetime.now(), duration=timedelta(minutes=5))
+
+
+def test_invalid_timezone_converted_to_utc() -> None:
+    """Test that non-UTC timezones are converted to UTC."""
+    start = datetime.now(timezone(timedelta(hours=1)))
+    period = DeliveryPeriod(start=start, duration=timedelta(minutes=5))
+
+    assert period.start.tzinfo == timezone.utc
```

Comment:
I wonder if we also should check if the time has been shifted according to the tz change.

Reply:
Can do but imo I don't think it really makes sense given the way the conversion is done

Reply:
How the conversion is done is implementation specific. The test doesn't know anything about that and the implementation might change in the future.

Reply:
Yeah fair point, I added a line now to check